gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GcsJavaClientWriteChannel.java

```diff
@@ -0,0 +1,179 @@
+package com.google.cloud.hadoop.gcsio;
+
+import static com.google.cloud.hadoop.gcsio.GoogleCloudStorageImpl.encodeMetadata;
+import static com.google.storage.v2.ServiceConstants.Values.MAX_WRITE_CHUNK_BYTES;
+
+import com.google.cloud.WriteChannel;
+import com.google.cloud.hadoop.util.AbstractGoogleAsyncWriteChannel;
+import com.google.cloud.hadoop.util.AsyncWriteChannelOptions;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.Storage.BlobWriteOption;
+import com.google.common.flogger.GoogleLogger;
+import com.google.common.io.ByteStreams;
+import com.google.protobuf.ByteString;
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+
+class GcsJavaClientWriteChannel extends AbstractGoogleAsyncWriteChannel<Boolean> {
+
+  protected static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  private final GoogleCloudStorageOptions storageOptions;
+  private final StorageResourceId resourceId;
+  private final AsyncWriteChannelOptions channelOptions;
+  private final Storage storage;
+  private WriteChannel writeChannel;
+  private final CreateObjectOptions createOptions;
+  private Boolean uploadSuccess = false;
+  // TODO: not supported as of now
+  // private final String requesterPaysProject;
+
+  public GcsJavaClientWriteChannel(
+      Storage storage,
+      GoogleCloudStorageOptions storageOptions,
+      StorageResourceId resourceId,
+      CreateObjectOptions createOptions,
+      ExecutorService uploadThreadPool) {
```

Comment:
I thought GCS client has the background upload etc. Is that not the case?

Reply:
Java-Client do provide a way to configure additional buffer and do manage streams to submit the provided chunks to gcs. But all of it is still a blocking call to gcs. I ma still utilising the IO pipe to reduce the backpressure.

Reply:
backpressure from where?

Why is using IO pipe better?

Reply:
BackPressure from Spark, that was the main reason it was introduced even for Json. To handled the back pressure we have introduced IO pipes.